airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/CatalogConverter.java

```diff
@@ -111,20 +162,62 @@ public static io.airbyte.protocol.models.ConfiguredAirbyteCatalog toProtocolKeep
    * @param catalog api catalog
    * @return protocol catalog
    */
-  public static io.airbyte.protocol.models.ConfiguredAirbyteCatalog toProtocol(final io.airbyte.api.model.generated.AirbyteCatalog catalog) {
+  public static io.airbyte.protocol.models.ConfiguredAirbyteCatalog toProtocol(final io.airbyte.api.model.generated.AirbyteCatalog catalog)
+      throws JsonValidationException {
+    final ArrayList<JsonValidationException> errors = new ArrayList<>();
     final List<io.airbyte.protocol.models.ConfiguredAirbyteStream> streams = catalog.getStreams()
         .stream()
         .filter(s -> s.getConfig().getSelected())
-        .map(s -> new io.airbyte.protocol.models.ConfiguredAirbyteStream()
-            .withStream(toProtocol(s.getStream()))
-            .withSyncMode(Enums.convertTo(s.getConfig().getSyncMode(), io.airbyte.protocol.models.SyncMode.class))
-            .withCursorField(s.getConfig().getCursorField())
-            .withDestinationSyncMode(Enums.convertTo(s.getConfig().getDestinationSyncMode(),
-                io.airbyte.protocol.models.DestinationSyncMode.class))
-            .withPrimaryKey(Optional.ofNullable(s.getConfig().getPrimaryKey()).orElse(Collections.emptyList())))
+        .map(s -> {
+          try {
+            return new io.airbyte.protocol.models.ConfiguredAirbyteStream()
+                .withStream(toProtocol(s.getStream(), s.getConfig()))
+                .withSyncMode(Enums.convertTo(s.getConfig().getSyncMode(), io.airbyte.protocol.models.SyncMode.class))
+                .withCursorField(s.getConfig().getCursorField())
+                .withDestinationSyncMode(Enums.convertTo(s.getConfig().getDestinationSyncMode(),
+                    io.airbyte.protocol.models.DestinationSyncMode.class))
+                .withPrimaryKey(Optional.ofNullable(s.getConfig().getPrimaryKey()).orElse(Collections.emptyList()));
+          } catch (JsonValidationException e) {
+            LOGGER.error("Error parsing catalog: {}", e);
+            errors.add(e);
+            return null;
+          }
+        })
         .collect(Collectors.toList());
+    if (!errors.isEmpty()) {
+      throw errors.get(0);
+    }
     return new io.airbyte.protocol.models.ConfiguredAirbyteCatalog()
         .withStreams(streams);
   }
 
+  /**
+   * Generate the map from StreamDescriptor to indicator of whether field selection is enabled for
+   * that stream.
+   *
+   * @param syncCatalog the catalog
+   * @return the map as a FieldSelectionData object
+   */
+  public static FieldSelectionData getFieldSelectionData(AirbyteCatalog syncCatalog) {
+    if (syncCatalog == null) {
+      return null;
+    }
+    final var fieldSelectionData = new FieldSelectionData();
+    syncCatalog.getStreams().stream().forEach((streamAndConfig) -> {
```

Comment:
nit: I've started preferring a `for (final StreamAndConfig streamAndConfig : syncCatalog.getStreams()) {` style instead of Java Streams for this sort of thing. I tend to only use Java Streams if I'm using collectors or maps, but totally fine to leave it as-is if you prefer

Reply:
I'm not so attached to Java Streams :)